[grid] Accommodate ability to specify sub-paths

krmahadevan opened the pull request:
Fixes: #10847

Sub-paths can now be specified for the below 
Components:
1. Hub
2. Standalone
3. Router

It can be specified via “--sub-path <actualPathGoesHere>”

Or via the Toml configuration file by specifying

[reverse_proxy]
sub-path = “<actualPathGoesHere>”

**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.


<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Describe your changes in detail -->

Sub-paths can now be specified for the below 
Components:
1. Hub
2. Standalone
3. Router

It can be specified via “--sub-path <actualPathGoesHere>”

Or via the Toml configuration file by specifying

[reverse_proxy]
sub-path = “<actualPathGoesHere>”


### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->

krmahadevan commented:
@diemol - I have only run `bazel test --cache_test_results=no java/test/org/openqa/selenium/grid/router:medium-tests`

diemol commented on java/src/org/openqa/selenium/grid/server/ReverseProxyFlags.java:
```diff
@@ -0,0 +1,45 @@
+import org.openqa.selenium.grid.config.HasRoles;
+import org.openqa.selenium.grid.config.Role;
+
+@AutoService(HasRoles.class)
+public class ReverseProxyFlags implements HasRoles {
+
+  @Parameter(
+    names = {"--sub-path"},
+    arity = 1,
+    description = "A sub-path that should be considered for all user facing routes on the Hub/Router/Standalone")
+  @ConfigValue(section = REVERSE_PROXY_SECTION, name = "sub-path", example = "my_company/selenium_grid")
+  public String subPath;
+
+  @Override
+  public Set<Role> getRoles() {
+    return Collections.singleton(ROUTER_ROLE);
+  }
```
This can be part of the `RouterFlags`, no need for a new class.

krmahadevan replied:
Doing this would basically mean that I would need to alter dependencies because it looks like both `Hub` and `Standalone` don't have this package exposed to them via bazel. Is that fine @diemol

krmahadevan replied:
Am seeing this when running the medium-tests from command line

```
# Execution platform: @local_config_platform//:host
java/src/org/openqa/selenium/grid/commands/Hub.java:44: error: package org.openqa.selenium.grid.router.httpd does not exist
import org.openqa.selenium.grid.router.httpd.RouterOptions;
```

I guess bazel is not happy here ?

krmahadevan replied:
Please ignore this. Figured out what needs to be done.

diemol commented on java/src/org/openqa/selenium/grid/server/ReverseProxyOptions.java:
```diff
@@ -0,0 +1,44 @@
+
+import org.openqa.selenium.grid.config.Config;
+
+public class ReverseProxyOptions {
+
+  static final String REVERSE_PROXY_SECTION = "reverse_proxy";
+
+  private final Config config;
+
+  public ReverseProxyOptions(Config config) {
+    this.config = config;
+  }
+
+  public String subPath() {
+    return config.get(REVERSE_PROXY_SECTION, "sub-path").map(prefix -> {
+      prefix = prefix.trim();
+      if (!prefix.startsWith("/")) {
+        prefix = "/" + prefix; //Prefix with a '/' if absent.
+      }
+      if (prefix.endsWith("/")) {
+        prefix = prefix.substring(0, prefix.length() -1); //Remove the trailing '/' if present.
+      }
+      return prefix;
+    }).orElse("");
+  }
```
You can create this as the `RouterOptions` if you wish.

diemol commented on java/src/org/openqa/selenium/grid/web/GridUiRoute.java:
```diff
@@ -65,4 +77,34 @@ public boolean matches(HttpRequest req) {
     return routes.execute(req);
   }
+
+  private static Routable[] uiRoutes(String prefix, Supplier<HttpHandler> handler) {
+    Routable alwaysAdd = Route.prefix("/ui").to(Route.matching(req -> true).to(handler));
+    if (prefix.isEmpty()) {
+      return new Routable[]{alwaysAdd};
+    }
+    return new Routable[]{
+      alwaysAdd,
+      Route.prefix(prefix + "/ui").to(Route.matching(req -> true).to(handler))
+    };
+  }
+
+  private static Routable[] consoleRoutes(String prefix, Supplier<HttpHandler> handler) {
+    Routable alwaysAdd = get("/grid/console").to(handler);
+    if (prefix.isEmpty()) {
+      return new Routable[]{alwaysAdd};
+    }
+    return new Routable[]{
+      alwaysAdd,
+      get(prefix + "/grid/console").to(handler)
+    };
+  }
+
+  private static Routable[] redirectRoutes(String prefix, Supplier<HttpHandler> handler ) {
+    if (prefix.isEmpty()) {
+      return new Routable[]{};
+    }
+    prefix = prefix.endsWith("/") ? prefix.substring(0, prefix.length() -1) : prefix;
+    return new Routable[]{get(prefix).to(handler)};
+  }
```
This could be a single method by adding a 3rd parameter, which is the route itself, right?

krmahadevan replied:
I didn't quite understand this recommendation @diemol Can you please help elaborate? These 3 methods are adding different types of routes. Earlier they were 3 individual lines doing that, but now since we need to handle the prefix as well, i extracted them out as 3 specific methods.

diemol replied:
But aren't these three methods doing almost the same? Just with different values for the Route?

krmahadevan replied:
* `uiRoutes()` works with `Route.prefix()`
* `consoleRoutes()` works with `Route.get()`
* `redirectRoutes()` is applicable only if there was a prefix that was specified.

Unifying them would basically mean that we create one method which internally does all these 3 activities. It would be the same as having 3 explicit methods which explicitly clarify their intent.

krmahadevan replied:
Please check now. I have refactored it and simplified the logic a bit more.

diemol reviewed (changes requested):
Thank you, @krmahadevan!

Left some comments so we can iterate.

yashcho commented:
Any update @diemol ?

diemol commented on java/src/org/openqa/selenium/grid/router/httpd/RouterOptions.java:
```diff
@@ -0,0 +1,44 @@
+
+public class RouterOptions {
+
+  static final String REVERSE_PROXY_SECTION = "reverse_proxy";
```
This can be called `NETWORK_SECTION`, naming it reverse proxy makes it very specific. It would then go together with the `relaxChecks` option already present in the `RouterFlags`.

krmahadevan replied:
@diemol - I didn't understand this part. 

Are you suggesting something like this ? 

```java
static final String NETWORK_SECTION = "network";
```

diemol replied:
Yes

krmahadevan commented:
@diemol - I have addressed the last review comment, rebased off of trunk and force pushed this branch. Please take a look. I will send a separate PR for this for addressing the documentation needs.

diemol reviewed (approved):
Thank you, @krmahadevan!

yashcho commented:
@diemol / @krmahadevan any idea in which release is it planned to rollout?

titusfortner commented:
Should be available in 4.8 (the latest release)

yashcho commented:
Thank you @titusfortner for information but I don't see it in release notes, even any idea on how it will work for docker-selenium?

diemol commented:
I will publish the release blog post today. What is missing is the documentation for this feature in the docs.

yashcho commented:
Thank you @diemol as I didn't work completely on creating this feature it would be difficult for me to write docs.

diemol commented:
Apologies if my comment was ambiguous, @yashcho. I meant, we at Selenium need to document the feature.

yashcho commented:
Understood, I would wait for the release note and feature doc.. :-)
